### ISSUE TITLE
Backport pr #1388 to v0.6

### ISSUE THREAD

**Issue by Thomasdezeeuw** [tokio-rs/mio#1390]
Pr: https://github.com/tokio-rs/mio/pull/1388
Issue: https://github.com/tokio-rs/mio/issues/1386

**Comment by Thomasdezeeuw** [comment]
Mio 0.6.x itself is fine, but the problem is in the net2 crate: https://github.com/deprecrated/net2-rs/issues/105.

**Comment by faern** [comment]
So when the `net2` PR is merged and `net2 0.2.36` is released. Let's bump `mio` so it says it requires **at least** `net2 = 0.2.36`? That way people will get it even if they later do `cargo update -p mio` in their projects.

Would be great if we could also do the same for `tokio`. Get a `tokio 0.2.23` patch release published that forces the use of this newer `mio`.

My own experience is that selectively upgrading some dependencies is somewhat common. And we would like this patch to propagate as fast as possible.

**Comment by Thomasdezeeuw** [comment]
> So when the `net2` PR is merged and `net2 0.2.36` is released. Let's bump `mio` so it says it requires **at least** `net2 = 0.2.36`? That way people will get it even if they later do `cargo update -p mio` in their projects.

Yes that sounds good.

> Would be great if we could also do the same for `tokio`. Get a `tokio 0.2.23` patch release published that forces the use of this newer `mio`.

I think we can do that, but I want some more prs merged before releasing Mio both 0.7.x and 0.6x.

**Comment by faern** [comment]
I'm thinking that maybe `mio 0.6` is not completely without guilt. Or I have made a mistake. Because I try to run the `mio` test suite under my toolchain built on https://github.com/rust-lang/rust/pull/78802. Without patching `net2` I get invalid pointer access (to be expected) but if I build it on top of the `net2` code in https://github.com/deprecrated/net2-rs/pull/106 I still get test failures.

For example `test_tcp::peek` fails on
```rust
poll.register(&s, Token(1), Ready::readable(), PollOpt::edge()).unwrap();
```
with error kind `AddrNetAvailable` and the message "The requested address is not valid in its context.".

I have not yet been able to figure out why. I'm not at all knowledgeable in the mio codebase really.

**Comment by faern** [comment]
I don't think it's the `net2` PR that is faulty. I have added tests to it where it prints out the `c::sockaddr` struct content as a byte slice. And with the PR code the results are identical to master for both IPv4 and IPv6 :shrug:

**Comment by faern** [comment]
Ah.. `miow` also does the same memory layout assumption: https://github.com/yoshuawuyts/miow/blob/c86373961b97eba692374435e93b81871af3f611/src/net.rs#L459-L470

I'll try to get PRs for that up during the weekend.

**Comment by faern** [comment]
This can be closed, since #1405 is merged and published.
